server/src/common/dto/xp-systems/xp-systems.dto.ts

```diff
@@ -0,0 +1,171 @@
+﻿import { ApiProperty, PickType } from '@nestjs/swagger';
+import { NestedDto } from '@lib/dto.lib';
+import { IsNumber } from 'class-validator';
+import { Prisma } from '@prisma/client';
+
+// Could implement RankXpParams and CosXpParams in xp-systems.interface.ts but class-validator requires
+// we give it this wacky structure.
+
+// TODO: bunch of this can be readonly?
+class CosXpTierScale {
+    [key: string]: Prisma.JsonValue;
+    @ApiProperty()
+    linear: number;
+
+    @ApiProperty()
+    staged: number;
+
+    @ApiProperty()
+    stages: number;
+
+    @ApiProperty()
+    bonus: number;
+}
+
+class CosXpUniqueTierScale
+    extends PickType(CosXpTierScale, ['linear', 'staged'] as const)
+    implements Record<string, Prisma.JsonValue>
+{
+    [key: string]: Prisma.JsonValue;
+}
+class CosXpRepeatTierScale extends CosXpTierScale {}
+class CosXpUnique {
+    [key: string]: Prisma.JsonObject;
+    @NestedDto(CosXpRepeatTierScale, { type: () => CosXpRepeatTierScale })
+    tierScale: CosXpUniqueTierScale;
+}
+class CosXpRepeat {
+    [key: string]: Prisma.JsonObject;
+    @NestedDto(CosXpRepeatTierScale, { type: () => CosXpRepeatTierScale })
+    tierScale: CosXpRepeatTierScale;
+}
+
+class CosXpCompletions {
+    [key: string]: Prisma.JsonObject;
+    @NestedDto(CosXpUnique, { type: () => CosXpUnique })
+    unique: CosXpUnique;
+
+    @NestedDto(CosXpRepeat, { type: () => CosXpRepeat })
+    repeat: CosXpRepeat;
+}
+class CosXpLevels {
+    [x: string]: Prisma.JsonValue;
+    @ApiProperty()
+    maxLevels: number;
+
+    @ApiProperty()
+    startingValue: number;
+
+    @ApiProperty()
+    linearScaleBaseIncrease: number;
+
+    @ApiProperty()
+    linearScaleInterval: number;
+
+    @ApiProperty()
+    linearScaleIntervalMultiplier: number;
+
+    @ApiProperty()
+    staticScaleStart: number;
+
+    @ApiProperty()
+    staticScaleBaseMultiplier: number;
+
+    @ApiProperty()
+    staticScaleInterval: number;
+
+    @ApiProperty()
+    staticScaleIntervalMultiplier: number;
+}
+
+class CosXpDto {
+    [key: string]: Prisma.JsonObject;
+    @NestedDto(CosXpCompletions, { type: () => CosXpCompletions })
```

Comment:
Is the explicit `type: () => ...` needed here? That should only be required for avoiding circular references in Swagger. Applies to several of these files